sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala

```diff
@@ -103,188 +85,247 @@ private object GpuRowToColumnConverter {
-      column: Int, builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Long =
+      column: Int, builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double =
       if (row.isNullAt(column)) {
         builder.appendNull()
-        0
+        4.125 // Offset and validity
       } else {
-        NotNullStringConverter.append(row, column, builder)
+        NotNullStringConverter.append(row, column, builder) + 4.125
       }
   }
 
-  private object NotNullStringConverter extends VariableWidthTypeConverter {
+  private object NotNullStringConverter extends TypeConverter {
     override def append(row: SpecializedGetters,
       column: Int,
-      builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Long = {
+      builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
       val bytes = row.getUTF8String(column).getBytes
       builder.appendUTF8String(bytes)
-      bytes.length
+      bytes.length + 4 // Includes offset
     }
   }
 
-  // ONLY supports Map(String, String)
-  private case object MapConverter
-    extends VariableWidthTypeConverter {
+  private[this] def mapConvert(
+      keyConverter: TypeConverter,
+      valueConverter: TypeConverter,
+      row: SpecializedGetters,
+      column: Int,
+      builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder) : Double = {
+    var ret = 0.0
+    val m = row.getMap(column)
+    val numElements = m.numElements()
+    val srcKeys = m.keyArray()
+    val srcValues = m.valueArray()
+    val structBuilder = builder.getChild(0)
+    val keyBuilder = structBuilder.getChild(0)
+    val valueBuilder = structBuilder.getChild(1)
+    for (i <- 0 until numElements) {
+      ret += keyConverter.append(srcKeys, i, keyBuilder)
+      ret += valueConverter.append(srcValues, i, valueBuilder)
+      structBuilder.endStruct()
+    }
+    builder.endList()
+    ret + 4
+  }
+
+  private case class MapConverter(
+      keyConverter: TypeConverter,
+      valueConverter: TypeConverter) extends TypeConverter {
     override def append(row: SpecializedGetters,
-      column: Int, builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Long = {
-      import scala.collection.JavaConverters._
-      val m = row.getMap(column)
-      val numElements = m.numElements()
-      val srcKeys = m.keyArray()
-      val srcValues = m.valueArray()
-      val listOfData = new ArrayBuffer[StructData](numElements)
-      (0 until srcKeys.numElements()).foreach { i =>
-        val values = new ArrayBuffer[AnyRef]()
-        values += new String(srcKeys.getUTF8String(i).getBytes)
-        values += new String(srcValues.getUTF8String(i).getBytes)
-        listOfData += new StructData(values.asJava)
+        column: Int, builder: ai.rapids.cudf.HostColumnVector.ColumnBuilder): Double = {
+      if (row.isNullAt(column)) {
+        builder.appendNull()
+        4.125
+      } else {
+        mapConvert(keyConverter, valueConverter, row, column, builder) + 0.125
```

Comment:
Some comments here will make it easy for newbies to follow along

Reply:
I assumed you wanted some help on the return values, but if you want other comments let me know and I'll add them.

Reply:
latest change looks good to me.